components/raft_log_engine/src/engine.rs

```diff
@@ -242,6 +242,10 @@ impl RaftEngineReadOnly for RaftLogEngine {
             .fetch_entries_to::<MessageExtTyped>(raft_group_id, begin, end, max_size, to)
             .map_err(transfer_error)
     }
+
+    fn get_all_entries_to(&self, _region_id: u64, _buf: &mut Vec<Entry>) -> Result<()> {
```

Comment:
/cc @tabokie PTAL, I'm not sure how to implement it correctly with raft engine.

Reply:
There is no shortcut here. Something like this:
```
if let (Some(first), Some(last)) = (self.0.first_index(id), self.0.last_index(id)) {
    buf.reserve(last - first + 1);
    self.fetch_entries_to(id, first, last + 1, None /*max_size*/, &mut buf)?;
}
```

Reply:
The thing is `get_all_entries_to` should not check if entries are contiguous. All it does is fetching all available entries in the region.

Reply:
Raft engine doesn't allow non-contiguous entries, so there is not extra check either.

Reply:
Why `first_index` is not reliable?

Reply:
For example, there are [4, 16] logs in a peer. The peer schedules raft log gc task delete [4, 8), and set first index to 8. While the gc task is pending in queue, the peer is deleted. So logs in range [8, 16] are deleted first, [4, 8) are deleted later.

If TiKV is shutdown when the task is in queue and the peer is not deleted, then after restart, its first index becomes 8. If it applies a snapshot, then it will delete logs in [8, 16] first, and then apply the snapshot. It may append logs further, for example, [20, 24]. And then delete [4, 8). So there is a time both [4, 8) and [20, 24] exist in the engine.

Reply:
`get_all_entries_to` is used to make sure no matter the order of executions, there will be no garbage left in the end.

Reply:
Oh, this won't happen with raft engine. Raft engine's `compact_to` ignores the `from` index and deletes all entries in `[0, to]`.

Reply:
Feel free to modify. It actually returns wrong result if there are multiple tasks of the same region.